lib/teleterm/services/remoteclientcache/remoteclientcache.go

```diff
@@ -0,0 +1,194 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package remoteclientcache
+
+import (
+	"context"
+	"sync"
+
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/singleflight"
+
+	"github.com/gravitational/teleport/lib/client"
+	"github.com/gravitational/teleport/lib/teleterm/api/uri"
+	"github.com/gravitational/teleport/lib/teleterm/clusters"
+)
+
+// CacheI defines an interface for proxy client cache.
+type CacheI interface {
+	// Get returns a proxy client from the cache if there is one,
+	// otherwise it dials the remote server.
+	Get(ctx context.Context, clusterURI uri.ResourceURI) (*client.ProxyClient, error)
+	// InvalidateForRootCluster closes and removes clients from the cache
+	// for the root cluster and its leaf clusters.
+	InvalidateForRootCluster(rootClusterURI uri.ResourceURI) error
+	// Close closes and removes all clients.
+	Close() error
+}
+
+// Cache stores remote clients keyed by cluster URI.
+// Safe for concurrent access.
+// Closes all clients and wipes the cache on Close.
+type Cache struct {
+	clusters.Resolver
+	mu  sync.Mutex
+	log *logrus.Entry
+	// clients keep mapping between cluster URI
+	// (both root and leaf) and proxy clients
+	clients map[uri.ResourceURI]*client.ProxyClient
+	// group prevents duplicate requests to create remote clients
+	// for a given cluster URI
+	group singleflight.Group
+}
+
+func New(c Config) *Cache {
+	return &Cache{
+		log:      c.Log,
+		clients:  make(map[uri.ResourceURI]*client.ProxyClient),
+		Resolver: c.Resolver,
+	}
+}
+
+func (c *Cache) Get(ctx context.Context, clusterURI uri.ResourceURI) (*client.ProxyClient, error) {
+	cltI, err, _ := c.group.Do(clusterURI.String(), func() (interface{}, error) {
+		cluster, clusterClient, err := c.ResolveCluster(clusterURI)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		// Check if we already have a cached client for this cluster and a valid cert.
+		// Theoretically, we don't have to check the validity
+		// because auth server will disconnect expired clients.
+		// However, by default, it isn't done immediately
+		// (until it is enabled with disconnect_expired_cert on cluster) and to keep
+		// the previous behavior, we need to check the cert.
+		//
+		// If the cert is expired, we will remove the client and make an attempt
+		// to connect to proxy which will fail with an appropriate error.
+		proxyClient := c.getFromCache(clusterURI)
+		if proxyClient != nil {
+			if cluster.Connected() {
+				return proxyClient, nil
+			}
+			err := c.InvalidateForRootCluster(clusterURI.GetRootClusterURI())
+			if err != nil {
+				c.log.WithError(err).Errorf("Failed to invalidate expired remote client for %q.", clusterURI)
+			}
+		}
```

Comment:
What's the benefit of invalidating the client on `Get`?

As it is, each call to get a client will incur the overhead of `c.ResolveCluster` (which reads from tsh home dir) in order to check if the cert is valid. But couldn't we make `Get` fairly simple and move the invalidation logic to places where we're certain that invalidation needs to happen?

Off the top of my head, `Logout` would need to remove the client from the cache while `Login` and `AssumeRole` would need to replace any old client with a new one.

Reply:
Actually, in the first version I didn't have that client invalidation here. It was something like this:
```go
if proxyClient := c.getFromCache(clusterURI); proxyClient != nil && cluster.Connected() {
	return proxyClient, nil
}
		
newProxyClient, err := clusterClient.ConnectToCluster(ctx)
if err != nil {
...
```
The previous client would be closed when adding a new client or by the remote server, but then I thought that it would be good to close the connection immediately and added that `InvalidateForRootCluster` call.

But in general, do we need checking if the cert is valid? Probably no, the auth server will terminate the session anyway.

Reply:
I agree, I don't think there's a lot of value in client-side checking of certificate expiration; only the auth server's point of view matters, after all.